src/bin/pg_autoctl/supervisor.c

```diff
@@ -0,0 +1,646 @@
+					 service->name, service->pid);
+		}
+		else
+		{
+			int idx = 0;
+
+			log_error("Failed to start service %s, "
+					  "stopping already started services and pg_autoctl",
+					  service->name);
+
+			for (idx = serviceIndex - 1; idx > 0; idx--)
+			{
+				if (kill(services[idx].pid, SIGQUIT) != 0)
+				{
+					log_error("Failed to send SIGQUIT to service %s with pid %d",
+							  services[idx].name, services[idx].pid);
+				}
+			}
+
+			/* we return false always, even if supervisor_stop is successful */
+			(void) supervisor_stop(pidfile);
+
+			return false;
+		}
+	}
+
+	/* now supervise sub-processes and implement retry strategy */
+	if (!supervisor_loop(start_pid, services, serviceCount, pidfile))
+	{
+		log_fatal("Something went wrong in sub-process supervision, "
+				  "stopping now. See above for details.");
+		success = false;
+	}
+
+	return supervisor_stop(pidfile) && success;
+}
+
+
+/*
+ * service_supervisor calls waitpid() in a loop until the sub processes that
+ * implement our main activities have stopped, and then it cleans-up the PID
+ * file.
+ */
+static bool
+supervisor_loop(pid_t start_pid,
+				Service services[],
+				int serviceCount,
+				const char *pidfile)
+{
+	int subprocessCount = serviceCount;
+	int stoppingLoopCounter = 0;
+	bool shutdownSequenceInProgress = false;
+	bool cleanExit = false;
+
+	/* wait until all subprocesses are done */
+	while (subprocessCount > 0)
+	{
+		pid_t pid;
+		int status;
+
+		/* Check that we still own our PID file, or quit now */
+		(void) check_pidfile(pidfile, start_pid);
+
+		/* If necessary, now is a good time to reload services */
+		if (asked_to_reload)
+		{
+			int serviceIndex = 0;
+
+			for (serviceIndex = 0; serviceIndex < serviceCount; serviceIndex++)
+			{
+				Service *service = &(services[serviceIndex]);
+
+				log_debug("Reloading pg_autoctl %s service", service->name);
+
+				(void) (*service->reloadFunction)(service);
+			}
+
+			/* reset our signal handling facility */
+			asked_to_reload = 0;
+		}
+
+		/* ignore errors */
+		pid = waitpid(-1, &status, WNOHANG);
+
+		switch (pid)
+		{
+			case -1:
+			{
+				if (errno == ECHILD)
+				{
+					/* no more childrens */
+					if (asked_to_stop || asked_to_stop_fast)
+					{
+						/* off we go */
+						log_info("Internal subprocesses are done, stopping");
+						return true;
+					}
+				}
+				else
+				{
+					log_fatal("Failed to call waitpid(): %m");
+					return false;
+				}
+			}
+
+			case 0:
+			{
+				/*
+				 * We're using WNOHANG, 0 means there are no stopped or exited
+				 * children, it's all good. It's the expected case when
+				 * everything is running smoothly, so enjoy and sleep for
+				 * awhile.
+				 */
+				if (asked_to_stop || asked_to_stop_fast)
+				{
+					cleanExit = true;
+
+					/*
+					 * Stop all the services.
+					 */
+					if (stoppingLoopCounter == 0)
+					{
+						(void) supervisor_stop_subprocesses(services,
+															subprocessCount);
```

Comment:
I think I see what you mean, and I'm not sure it would be of much use. Will have to look at the source again with the other PRs to see if we would then be able to have the same function everywhere just calling the new API, but I think we have different local variables specific to each service that we want to use in each place anyway.

Reply:
The idea of `asked_to_stop_fast` is just that we don't finish the current keeper loop and quit as soon as possible, and `asked_to_stop` allows for finishing current business first. I believe we still have that difference in the way the main keeper look reacts to signals. That might be the only place where we process those two signal variables differently too.

Reply:
> The idea of asked_to_stop_fast is just that we don't finish the current keeper loop and quit as soon as possible, and asked_to_stop allows for finishing current business first.

Without `stopFastFunction` this would not be possible then. Since the supervisor process will swallow the signal and send the same signal to the keeper process no matter what.

Reply:
The refactoring I just completed now relays the `asked_to_stop_fast` signal to sub-processes:

``` c
	int signal = asked_to_stop_fast ? SIGINT : SIGTERM;
```